### PR TITLE
ENH: Holt-Winters simulations

### PR DESCRIPTION
**Pull request by s-scherrer**
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

Implementation of `HoltWintersResults.simulate` as discussed in #6359, as a first step towards prediction intervals.

This is still work in progress, but I would like to get some feedback about the general direction/coding style/implementation details.

I didn't manage to reproduce the results from R's forecast package exactly in the current test, but the difference is low. How exact does it have to be, @ChadFulton?

### CODE REVIEW

**Comment by s-scherrer** [comment]
>     * [``](https://github.com/statsmodels/statsmodels/blob/aed6a666980fbb78d146927f89b299b53f7dd3ba/statsmodels/tsa/holtwinters.py):

This is on purpose to follow the mathematical notation in FPP2. Should I change this to somewhat else?

File: statsmodels/tsa/holtwinters.py [hunk 1]
```diff
@@ -430,6 +430,227 @@ def _fmt(x):
...
+            phi = 1
+        if seasonal is None:
+            s[:,:] = neutral_s
+        if not damped:
+            phi = 1
+
+        if error == 'add':
+            # estimate error variance from residuals
+            sigma = np.sqrt(np.sum(self.resid**2) / (len(self.resid) - n_params))
+            eps = np.random.randn(steps, nsim) * sigma
```
Comment (bashtage) [inline statsmodels/tsa/holtwinters.py]: Need to let a user (optionally) pass in a RandomState rather than using the global generator.

**Comment by bashtage** [comment]
You need to check if matplotib is available, and if not skip that part of the test.  Ideally split the matplotlib and non-matpltlib then you can `@pytest.mark.matplotlib` the test that uses MPL to skip it when not available.

File: statsmodels/tsa/tests/test_holtwinters.py [hunk 2]
```diff
@@ -358,6 +378,56 @@ def test_hw_seasonal_buggy(self):
...
+            self.austourists, seasonal_periods=4, trend="add", seasonal="mul",
+            damped=False
+        ).fit()
+
+        np.random.seed(0)
+        nsim = 100
+        steps = 20
+        sim = fit.simulate(steps, nsim=nsim, error="mul")
+
+        import matplotlib.pyplot as plt
```
Comment (bashtage) [inline statsmodels/tsa/tests/test_holtwinters.py]: Make a plotting only test.

File: statsmodels/tsa/holtwinters.py (hunk 1 above)
Comment (ChadFulton) [reply statsmodels/tsa/holtwinters.py]: In addition to supporting RandomState, I think we should also allow a few other cases:

1. User passes in the `eps` they want to use directly (helpful for replication / unit testing)
2. User passes in a scipy distribution function they want to use (allows non-Gaussian errors)
3. (Optional, we can leave this as a TODO) User specifies bootstrapping, in which case we re-sample the in-sample errors.

File: statsmodels/tsa/tests/test_holtwinters.py (hunk 2 above)
Comment (s-scherrer) [reply statsmodels/tsa/tests/test_holtwinters.py]: Sorry, I should have removed this before. This was just for myself as a first visual sanity check, not a real test. I removed it now.

**Comment by s-scherrer** [comment]
I made the following changes:
* Added keyword arguments for passing a random state/seed, a scipy distribution function, fixed errors, or bootstrap
* Moved tests to the end of the test file, using pytest fixtures for the data instead of the test class
* The tests set up a ``HoltWintersResults`` object with exactly the same state as is obtained from R's forecast package for all stable models (the ones that forecast allows). With this, predictions match up to 5 decimals.
* Added a test that checks whether the different keyword options are accepted and run (no comparison of outputs)

File: statsmodels/tsa/tests/test_holtwinters.py (hunk 2 above)
Comment (s-scherrer) [reply statsmodels/tsa/tests/test_holtwinters.py]: Sorry, I forgot to remove this, this was not thought of as a real test, but a first visual sanity check for me. I removed it now.

File: statsmodels/tsa/holtwinters.py (hunk 1 above)
Comment (s-scherrer) [reply statsmodels/tsa/holtwinters.py]: I implemented all these with two new keyword arguments:
* `seed` accepts a `RandomState` or an integer
* `random_errors` can be None, an array of values that are used directly, a scipy distribution function, or `"bootstrap"`.

**Comment by ChadFulton** [comment]
Thanks for this! I should have time for a review this weekend.

**Comment by s-scherrer** [comment]
@ChadFulton I just found that there's still an issue when the fit was performed with `use_boxcox=True`. I am working on it.

File: statsmodels/tsa/holtwinters.py [hunk 3]
```diff
@@ -944,12 +1250,12 @@ def _predict(self, h=None, smoothing_level=None, smoothing_slope=None,
         level = lvls[1:nobs + 1].copy()
         if use_boxcox or use_boxcox == 'log' or isinstance(use_boxcox, float):
             fitted = inv_boxcox(fitted, lamda)
-            level = inv_boxcox(level, lamda)
```
Comment (s-scherrer) [inline statsmodels/tsa/holtwinters.py]: I don't really understand why `level`, `slope` and `season` are "back-transformed" here. For the simulations I need the values in the transformed space, therefore I commented them for now.

The tests are still running fine, does this mean we can remove this?

Comment (ChadFulton) [reply statsmodels/tsa/holtwinters.py]: That is interesting. I can't see why the back-transformation would be sensible, and I notice that R's forecast package does not back-transform the states.

The only question here is about backwards compatibility - if people are relying on the current behavior. But it looks like the current behavior might be best classified as a bug?

File: statsmodels/tsa/holtwinters.py [hunk 4]
```diff
@@ -430,6 +433,309 @@ def _fmt(x):
...
+            neutral_s = 0
+
+        # set initial values
+        # (notation as in https://otexts.com/fpp2/ets.html)
+        y = np.empty((steps, nsim))
+        lvl = np.empty((steps + 1, nsim)) # lvl instead of l because of E741
+        b = np.empty((steps + 1, nsim))
+        s = np.empty((steps + m, nsim))
+        # the following uses python's index wrapping
+        lvl[-1,:] = self.level[start]
```
Comment (ChadFulton) [inline statsmodels/tsa/holtwinters.py]: space after comma

File: statsmodels/tsa/holtwinters.py [hunk 5]
```diff
@@ -430,6 +433,309 @@ def _fmt(x):
...
+        b[-1,:] = self.slope[start]
+        if start == -1:
+            s[-m:,:] = np.tile(self.season[-m:], (nsim, 1)).T
+        elif 0 <= start and start < m:
+            initial_seasons = self.params["initial_seasons"]
+            _s = np.concatenate((initial_seasons[start+1:],
```
Comment (ChadFulton) [inline statsmodels/tsa/holtwinters.py]: space between operators like `+`

**Comment by ChadFulton** [comment]
This is looking great, thanks! I just have a couple of comments.

The only other thing is that the updates should be cleaned up to match up with PEP8, I noted a couple of them (like spaces after commas, or spaces between operators like `+`).

File: statsmodels/tsa/holtwinters.py [hunk 6]
```diff
@@ -430,6 +433,309 @@ def _fmt(x):
...
+    def simulate(self, steps, start=-1, nsim=1, error="add",
+                 random_errors=None, random_state=None):
+        r"""
+        Random simulations using the state space formulation.
+
+        Parameters
+        ----------
+        steps : int
+            The number of simulation steps.
+        start : int, optional
```
[the remark this answers is not in the document]
Comment (s-scherrer) [reply statsmodels/tsa/holtwinters.py:445]: Done, I very liberally copied the documentation and code regarding `anchor` from `MLEModel.simulate`.

File: statsmodels/tsa/holtwinters.py (hunk 3 above)
Comment (s-scherrer) [reply statsmodels/tsa/holtwinters.py:947]: Yes, I would also classifiy this as a bug. 
What would be the best way to handle this? I could, for a start, store the original values in a private attribute and use these for the simulations. Then we could open a separate issue and PR to discuss backwards compability issues.

Comment (ChadFulton) [reply statsmodels/tsa/holtwinters.py:947]: What you suggest is definitely the ideal way forward, but a little annoying for you. If you don't mind doing that, that would be great.

**Comment by ChadFulton** [comment]
I think this is ready to merge, except for updating the back-transformation to store the original values. If you want to do that, we can merge directly afterwards, assuming CI is still green.

Thanks again!

**Comment by ChadFulton** [comment]
This is great, thanks very much for this PR.

**Comment by s-scherrer** [comment]
Thanks a lot for your assistance. I had a lot of fun and learned quite a bit.

**Pull request merged**